### PR TITLE
Fix for extracting linux binaries from container image.

### PR DESCRIPTION
**Pull request by cpakulski**
To avoid hardcoding info about container layering, the following logic has been added for _linux_ platform:
- list all the files in the container image (_car list_ with _-vv_)
- capture 5 lines before envoy binary (usr/local/bin/envoy) is listed. This should contain _CreatedBy_ phase. 
- get the last occurrence of _CreatedBy_ phase. From that line extract the command used to create the layer.
- that command is used as pattern in _car extract_.

Fixes: #34 (partially) 
Signed-off-by: Christoph Pakulski <christoph@tetrate.io>

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
doing it this way is more complicated and inefficient than removing the `--created-by-pattern` entirely. I'd recommend that instead if you feel strongly about no maintenance.

File: bin/car_envoy.sh [hunk 1]
```diff
@@ -102,15 +103,21 @@ darwin) # https://github.com/Homebrew/homebrew-core/blob/master/Formula/envoy.rb
...
     files="$files usr/local/bin/envoy.dwp"
   else
     reference=envoyproxy/envoy:${version}
   fi
 
-  ${car} --created-by-pattern 'ADD linux' --strip-components 2 -qf "${reference}" ${files}
+  # List all files and layers. Look for $envoy_binary. Capture last 5 lines before $envoy_binary and find
+  # the last "CreatedBy" phase. From the line containing "CreatedBy" get the command used to add the layer.
+  # That command will be used as $pattern for extract command.
+  pattern=$(${car_base} -t -vv --reference "${reference}" | grep -B 5 "${envoy_binary}" | grep "CreatedBy" | tail -1 | awk '{print $2 " " $3}')
```
Comment (codefromthecrypt) [inline bin/car_envoy.sh:118]: ex this scans the all layers, especially as it doesn't use `-q` it scans all layers, not just the ones that include the file desires.

File: bin/car_envoy.sh [hunk 2]
```diff
@@ -102,15 +103,21 @@ darwin) # https://github.com/Homebrew/homebrew-core/blob/master/Formula/envoy.rb
...
     reference=envoyproxy/envoy:${version}
   fi
 
-  ${car} --created-by-pattern 'ADD linux' --strip-components 2 -qf "${reference}" ${files}
+  # List all files and layers. Look for $envoy_binary. Capture last 5 lines before $envoy_binary and find
+  # the last "CreatedBy" phase. From the line containing "CreatedBy" get the command used to add the layer.
+  # That command will be used as $pattern for extract command.
+  pattern=$(${car_base} -t -vv --reference "${reference}" | grep -B 5 "${envoy_binary}" | grep "CreatedBy" | tail -1 | awk '{print $2 " " $3}') 
+
+  ${car} --created-by-pattern "${pattern}" -qf "${reference}" --strip-components 2 ${files}
```
Comment (codefromthecrypt) [inline bin/car_envoy.sh:120]: now this goes right back to the layer the above command stopped at. there's no point in executing this twice. Remove all the changes above for the below if you prefer to have the job scan through unrelated layers instead of choosing them.

```suggestion
  ${car} -qf "${reference}" --strip-components 2 ${files}
```

**Review by codefromthecrypt — CHANGES_REQUESTED** [review]
here's a better way. one line change instead of several!

**Comment by codefromthecrypt** [comment]
the "not selecting layers approach" (aka just remove the `--created-by-pattern`) is fine for the main envoy image. You can verify in `docker history` the layers prior to the one containing the binary are tiny, and particularly when its -q flag is retained `car` will stop scanning once it finds the binaries. This is also fine for the debug image because the huge layer is actually the one with the binaries in it.

```bash
$ docker history envoyproxy/envoy:v1.22.0
IMAGE          CREATED       CREATED BY                                      SIZE      COMMENT
bfb5e3005bf8   3 weeks ago   CMD ["envoy" "-c" "/etc/envoy/envoy.yaml"]      0B        buildkit.dockerfile.v0
<missing>      3 weeks ago   ENTRYPOINT ["/docker-entrypoint.sh"]            0B        buildkit.dockerfile.v0
<missing>      3 weeks ago   EXPOSE map[10000/tcp:{}]                        0B        buildkit.dockerfile.v0
<missing>      3 weeks ago   RUN /bin/sh -c adduser --group --system envo…   44.2kB    buildkit.dockerfile.v0
<missing>      3 weeks ago   COPY /docker-entrypoint.sh / # buildkit         750B      buildkit.dockerfile.v0
<missing>      3 weeks ago   COPY /etc/envoy/envoy.yaml /etc/envoy/envoy.…   1.82kB    buildkit.dockerfile.v0
<missing>      3 weeks ago   COPY /usr/local/bin/su-exec /usr/local/bin/ …   19.1kB    buildkit.dockerfile.v0
<missing>      3 weeks ago   COPY /usr/local/bin/envoy* /usr/local/bin/ #…   59.4MB    buildkit.dockerfile.v0
<missing>      3 weeks ago   RUN /bin/sh -c mkdir -p /etc/envoy # buildkit   0B        buildkit.dockerfile.v0
<missing>      3 weeks ago   RUN /bin/sh -c apt-get update && apt-get upg…   6.83MB    buildkit.dockerfile.v0
<missing>      4 weeks ago   /bin/sh -c #(nop)  CMD ["bash"]                 0B        
<missing>      4 weeks ago   /bin/sh -c #(nop) ADD file:8a4ddbd462c1dd201…   63.2MB    
```

In other words, the above shows why avoiding selection of layers is fine on linux.

PS. This isn't ideal for the windows image because there are huge layers in front of the envoy binary. In any case, this drift didn't effect windows and also even if drift affects windows it may change other things, such as the path to the binary.

**Comment by cpakulski** [comment]
Thanks for reviewing and additional info about the layering. I will change it for linux to run without selecting the layer.
The other option would be to have a list of layer names and try them one by one. If they do not exist, issue a warning and extract the image without selecting the layer. Possible improvement for the future.

**Review by codefromthecrypt — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
thanks for the attention in learning this stuff @cpakulski
